File: src/osscodeiq/detectors/config/properties_detector.py
"""Properties file detector for Java .properties and Spring configuration files."""

from __future__ import annotations

from osscodeiq.detectors.base import DetectorContext, DetectorResult
from osscodeiq.models.graph import (
    EdgeKind,
    GraphEdge,
    GraphNode,
    NodeKind,
    SourceLocation,
)

_DB_KEYWORDS = {"url", "jdbc", "datasource"}
_MAX_KEYS = 200


class PropertiesDetector:
    """Detects property keys, Spring config markers, and database connections from .properties files."""

    name: str = "properties"
    supported_languages: tuple[str, ...] = ("properties",)

    def detect(self, ctx: DetectorContext) -> DetectorResult:
        result = DetectorResult()

        # parsed_data expected: {"type": "properties", "file": path, "data": {"key": "value", ...}}
        if not isinstance(ctx.parsed_data, dict):
            return result

        if ctx.parsed_data.get("type") != "properties":
            return result

        data = ctx.parsed_data.get("data")
        if not isinstance(data, dict):
            return result

        filepath = ctx.file_path
        file_id = f"props:{filepath}"

        # CONFIG_FILE node
        result.nodes.append(GraphNode(
            id=file_id,
            kind=NodeKind.CONFIG_FILE,
            label=filepath,
            fqn=filepath,
            module=ctx.module_name,
            location=SourceLocation(
                file_path=filepath,
                line_start=1,
            ),
            properties={"format": "properties"},
        ))

        # Process keys (limit to avoid node explosion)
        keys = list(data.items())[:_MAX_KEYS]

        for key, value in keys:
            if not isinstance(key, str):
                continue

            key_lower = key.lower()
            key_id = f"props:{filepath}:{key}"

            # Detect DB connection properties
            is_db = any(kw in key_lower for kw in _DB_KEYWORDS)

            if is_db:
                props: dict[str, object] = {"key": key}
                if isinstance(value, str):
                    props["value"] = value

                result.nodes.append(GraphNode(
                    id=key_id,
                    kind=NodeKind.DATABASE_CONNECTION,
                    label=key,
                    fqn=f"{filepath}:{key}",
                    module=ctx.module_name,
                    location=SourceLocation(file_path=filepath),
                    properties=props,
                ))
            else:
                props = {"key": key}
                if isinstance(value, str):
                    props["value"] = value

                # Detect Spring config
                if key.startswith("spring."):
                    props["spring_config"] = True

                result.nodes.append(GraphNode(
                    id=key_id,
                    kind=NodeKind.CONFIG_KEY,
                    label=key,
                    fqn=f"{filepath}:{key}",
                    module=ctx.module_name,
                    location=SourceLocation(file_path=filepath),
                    properties=props,
                ))

            result.edges.append(GraphEdge(
                source=file_id,
                target=key_id,
                kind=EdgeKind.CONTAINS,
                label=f"{filepath} contains {key}",
            ))

        return result

File: src/main/java/io/github/randomcodespace/iq/detector/config/PropertiesDetector.java
package io.github.randomcodespace.iq.detector.config;

import io.github.randomcodespace.iq.detector.AbstractStructuredDetector;
import io.github.randomcodespace.iq.detector.DetectorContext;
import io.github.randomcodespace.iq.detector.DetectorResult;
import io.github.randomcodespace.iq.model.CodeEdge;
import io.github.randomcodespace.iq.model.CodeNode;
import io.github.randomcodespace.iq.model.EdgeKind;
import io.github.randomcodespace.iq.model.NodeKind;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import io.github.randomcodespace.iq.detector.DetectorInfo;
import io.github.randomcodespace.iq.detector.ParserType;

/**
 * Detects property keys, Spring config markers, and database connections from .properties files.
 */
@DetectorInfo(
    name = "properties",
    category = "config",
    description = "Detects Java properties files and database connection strings",
    parser = ParserType.STRUCTURED,
    languages = {"properties"},
    nodeKinds = {NodeKind.CONFIG_FILE, NodeKind.CONFIG_KEY, NodeKind.DATABASE_CONNECTION},
    edgeKinds = {EdgeKind.CONTAINS},
    properties = {"url"}
)
@Component
public class PropertiesDetector extends AbstractStructuredDetector {

    private static final Set<String> DB_KEYWORDS = Set.of("url", "jdbc", "datasource");
    private static final int MAX_KEYS = 200;

    @Override
    public String getName() {
        return "properties";
    }

    @Override
    public Set<String> getSupportedLanguages() {
        return Set.of("properties");
    }

    @Override
    public DetectorResult detect(DetectorContext ctx) {
        Object parsedData = ctx.parsedData();
        if (parsedData == null) return DetectorResult.empty();

        Map<String, Object> pd = asMap(parsedData);
        if (!"properties".equals(getString(pd, "type"))) {
            return DetectorResult.empty();
        }

        Map<String, Object> data = getMap(pd, "data");
        if (data.isEmpty()) return DetectorResult.empty();

        String filepath = ctx.filePath();
        String fileId = "props:" + filepath;
        List<CodeNode> nodes = new ArrayList<>();
        List<CodeEdge> edges = new ArrayList<>();

        // CONFIG_FILE node
        CodeNode fileNode = new CodeNode(fileId, NodeKind.CONFIG_FILE, filepath);
        fileNode.setFqn(filepath);
        fileNode.setModule(ctx.moduleName());
        fileNode.setFilePath(filepath);
        fileNode.setLineStart(1);
        fileNode.setProperties(Map.of("format", "properties"));
        nodes.add(fileNode);

        // Process keys (limit to avoid node explosion)
        int count = 0;
        for (var entry : data.entrySet()) {
            if (count >= MAX_KEYS) break;
            String key = entry.getKey();
            Object value = entry.getValue();

            String keyLower = key.toLowerCase();
            String keyId = "props:" + filepath + ":" + key;

            boolean isDb = DB_KEYWORDS.stream().anyMatch(keyLower::contains);

            Map<String, Object> props = new HashMap<>();
            props.put("key", key);
            if (value instanceof String s) {
                props.put("value", s);
            }

            if (isDb) {
                CodeNode keyNode = new CodeNode(keyId, NodeKind.DATABASE_CONNECTION, key);
                keyNode.setFqn(filepath + ":" + key);
                keyNode.setModule(ctx.moduleName());
                keyNode.setFilePath(filepath);
                keyNode.setProperties(props);
                nodes.add(keyNode);
            } else {
                if (key.startsWith("spring.")) {
                    props.put("spring_config", true);
                }
                CodeNode keyNode = new CodeNode(keyId, NodeKind.CONFIG_KEY, key);
                keyNode.setFqn(filepath + ":" + key);
                keyNode.setModule(ctx.moduleName());
                keyNode.setFilePath(filepath);
                keyNode.setProperties(props);
                nodes.add(keyNode);
            }

            CodeEdge edge = new CodeEdge();
            edge.setId(fileId + "->" + keyId);
            edge.setKind(EdgeKind.CONTAINS);
            edge.setSourceId(fileId);
            edge.setTarget(new CodeNode(keyId, null, null));
            edges.add(edge);

            count++;
        }

        return DetectorResult.of(nodes, edges);
    }
}
